File: CodingInterviewsPythonSolutions/Binary Search Tree Iterator.py
class BSTIterator:

    def __init__(self, root: Optional[TreeNode]):
        self.stk = []
        self.addLeftChain(root)

    def addLeftChain(self, root):
        while root:
            self.stk.append(root)
            root = root.left

    def next(self) -> int:
        t = self.stk.pop()
        if t.right:
            self.addLeftChain(t.right)
        return t.val

    def hasNext(self) -> bool:
        return len(self.stk) > 0

File: CodingInterviewsJavaSolutions/Binary Search Tree Iterator.java
class BSTIterator {
    Stack<TreeNode> stk;

    public BSTIterator(TreeNode root) {
        stk = new Stack<>();
        addLeftChain(root);
    }
    
    private void addLeftChain(TreeNode root) {
        while (root != null) {
            stk.push(root);
            root = root.left;
        }
    }
    
    public int next() {
        TreeNode t = stk.pop();
        if (t.right != null) {
            addLeftChain(t.right);
        }
        return t.val;
    }
    
    public boolean hasNext() {
        return !stk.isEmpty();
    }
}
